### ISSUE TITLE
ERROR: One or more modules listed in cowboy.app are not present in ebin/*.beam

### ISSUE THREAD

**Issue by youthy** [joearms/ezwebframe#16]
hello,joe. I just run `make` then occurred the error above.my erl version is 18.

**Comment by lostcolony** [comment]
Hi Youthy. Loic is currently working on cowboy's master branch and so it doesn't seem to be stable. 

To get past that, in the rebar.config file, change where the cowboy dependency is selected as 'master', to instead be the latest, 1.0.2. I.e., the file should be changed to contain -

{deps, [
  {cowboy, ".*", {git, "git://github.com/extend/cowboy.git", "1.0.2"}}
]}.

Then delete your deps directory, and rerun make. That ~should~ move you past this issue (at least, when I checked out the code, ran make, the issues I got were resolved by following those steps).

**Comment by youthy** [comment]
Hi, lostcolony.
I really appreciate your help. That works. Thks.
But when I enter page `localhost:1456` the server made some crashes. I will find out.;-)

**Comment by lostcolony** [comment]
Beh, it looks like the last tweaks Joe made to ezwebframe were based on a version of master that doesn't exist, or so I assume from the breakages I'm seeing. I'm not sure what the right combination of ezwebframe commit <-> cowboy version will work correctly, or if the issues you saw vs the ones I saw were different.

Joe probably ought to peg a specific version of cowboy, so that Loic's (breaking) changes to master leave this unaffected.

**Comment by youthy** [comment]
the error prints like this

```
=ERROR REPORT==== 23-Jul-2015::11:33:36 ===
Error in process <0.151.0> with exit value:
{[{reason,undef},
  {mfa,{ezwebframe,init,3}},
  {stacktrace,[{ezwebframe,init,
                           [{tcp,http},
                            {http_req,#Port<0.1561>,ranch_tcp,keepalive,
                                      <0.151.0>,<<"GET">>,'HTTP/1.1',
                                      {{127,0,0,1},48746},
                                      <<"localhost">>,undefined,1456,<<"/">>,
                                      undefined,<<>>,undefined,[],
                                      [{<<"host">>,<<"localhost:1456">>},
                                       {<<"connection">>,<<"keep-alive">>},
                                       {<<"accept">>,
                                        <<"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8">>},
                                       {<<"user-agent">>,
                                        <<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36">>},
                                       {<<"accept-encoding">>,
                                        <<"gzip, deflate, sdch">>},
                                       {<<"accept-language">>,
                                        <<"zh-CN,zh;q=0.8,en;q=0.6,ja;q=0.4,zh-TW;q=0.2">>}],
                                      [{<<"connection">>,[<<"keep-alive">>]}],
                                      undefined,[],waiting,<<>>,undefined,
                                      false,waiting,[],<<>>,undefined},
                            {env,#Fun<ezwebframe_demos.0.106014447>}],
                           []},
               {cowboy_handler,handler_init,4,
                               [{file,"src/cowboy_handler.erl"},{line,64}]},
               {cowboy_protocol,execute,4,
                                [{file,"src/cowboy_protocol.erl"},
                                 {line,442}]}]},
  {req,[{socket,#Port<0.1561>},
        {transport,ranch_tcp},
        {connection,keepalive},
        {pid,<0.151.0>},
        {method,<<"GET">>},
        {version,'HTTP/1.1'},
        {peer,{{127,0,0,1},48746}},
        {host,<<"localhost">>},
        {host_info,undefined},
        {port,1456},
        {path,<<"/">>},
        {path_info,undefined},
        {qs,<<>>},
        {qs_vals,undefined},
        {bindings,[]},
        {headers,[{<<"host">>,<<"localhost:1456">>},
                  {<<"connection">>,<<"keep-alive">>},
                  {<<"accept">>,
                   <<"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8">>},
                  {<<"user-agent">>,
                   <<"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36">>},
                  {<<"accept-encoding">>,<<"gzip, deflate, sdch">>},
                  {<<"accept-language">>,
                   <<"zh-CN,zh;q=0.8,en;q=0.6,ja;q=0.4,zh-TW;q=0.2">>}]},
        {p_headers,[{<<"connection">>,[<<"keep-alive">>]}]},
        {cookies,undefined},
        {meta,[]},
        {body_state,waiting},
        {buffer,<<>>},
        {multipart,undefined},
        {resp_compress,false},
        {resp_state,waiting},
        {resp_headers,[]},
        {resp_body,<<>>},
        {onresponse,undefined}]},
  {opts,{env,#Fun<ezwebframe_demos.0.106014447>}}],
 [{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,442}]}]}
```

**Comment by lostcolony** [comment]
Right. There was a change in cowboy to specify what transport as part of
the init callback. That is, cowboy handlers went from having an init/2
function, to an init/3 function. I tried modifying/adding and delegating
the function signature, then chased down the next error (the fact various
cowboy return types went from just returning Response, to {Response,
Request}), and that turned into a bit of a rabbit hole. I got the static
pages being served properly, but they'd still error, and the websocket
stuff wasn't working. I'm not that familiar with the ezwebframe code base
though, nor the latest cowboy versions (my last project was at 0.9.0 I
think). Hopefully Joe will have a chance to take a look at it (and peg a
specific cowboy version in rebar.config as he does). But it looks like it
may take a bit of work.
